internal/cni-server/cni-plugin.go

```diff
@@ -319,38 +325,126 @@ func skipListening(serviceMeshMode string, pid string) bool {
+	if err != nil {
+		log.Errorf("get iface error: %v", err)
+		return err
+	}
+	rtnl, err := tc.Open(&tc.Config{})
+	if err != nil {
+		log.Errorf("open rtnl error: %v", err)
+		return err
+	}
+	defer func() {
+		if err := rtnl.Close(); err != nil {
+			log.Errorf("could not close rtnetlink socket: %v\n", err)
+		}
+	}()
+	qdiscs, err := rtnl.Qdisc().Get()
+	if err != nil {
+		log.Errorf("get qdisc error: %v", err)
+		return err
+	}
+	find := false
+	for _, qdisc := range qdiscs {
+		if qdisc.Kind == "clsact" && qdisc.Ifindex == uint32(iface.Index) {
+			find = true
+			break
+		}
+	}
+	if !find {
+		// init clasact if not exists
+		err := rtnl.Qdisc().Add(&tc.Object{
+			Msg: tc.Msg{
+				Family:  unix.AF_UNSPEC,
+				Ifindex: uint32(iface.Index),
+				Handle:  core.BuildHandle(0xFFFF, 0x0000),
+				Parent:  tc.HandleIngress,
+			},
+			Attribute: tc.Attribute{
+				Kind: "clsact",
+			},
+		})
+		if err != nil {
+			return err
+		}
+	}
+	ing := ebpfs.GetTCIngressProg()
+	if ing == nil {
+		return fmt.Errorf("can not get ingress prog")
 	}
 
-	obj := "bpf/mb_tc.o"
+	filter := tc.Object{
+		Msg: tc.Msg{
+			Family:  unix.AF_UNSPEC,
+			Ifindex: uint32(iface.Index),
+			// Handle:  0,
+			Parent: 0xFFFFFFF2, // ingress
+			Info: core.BuildHandle(
+				66,     // prio
+				0x0300, // protocol
+			),
+		},
+		Attribute: tc.Attribute{
+			Kind: "bpf",
+			BPF: &tc.Bpf{
+				FD:    uint32Ptr(uint32(ing.FD())),
+				Name:  stringPtr("mb_tc_ingress"),
+				Flags: uint32Ptr(0x1),
+			},
+		},
+	}
+	if err := rtnl.Filter().Add(&filter); err != nil {
+		return err
+	}
+	egress := ebpfs.GetTCEgressProg()
+	if ing == nil {
+		return fmt.Errorf("can not get ingress prog")
+	}
 
-	cmd = exec.Command("sh", "-c", fmt.Sprintf("tc filter add prio 66 dev %s ingress bpf da obj %s sec classifier_ingress", dev, obj))
-	err = cmd.Run()
-	if code := cmd.ProcessState.ExitCode(); code != 0 || err != nil {
-		return fmt.Errorf("failed to attach tc(ingress) to %s, unexpected exit code: %d, err: %v", dev, code, err)
+	filter = tc.Object{
+		Msg: tc.Msg{
+			Family:  unix.AF_UNSPEC,
+			Ifindex: uint32(iface.Index),
+			// Handle:  0,
+			Parent: 0xFFFFFFF3, // egress
+			Info: core.BuildHandle(
+				66,     // prio
+				0x0300, // protocol
+			),
+		},
+		Attribute: tc.Attribute{
+			Kind: "bpf",
+			BPF: &tc.Bpf{
+				FD:    uint32Ptr(uint32(egress.FD())),
+				Name:  stringPtr("mb_tc_egress"),
+				Flags: uint32Ptr(0x1),
+			},
+		},
 	}
-	cmd = exec.Command("sh", "-c", fmt.Sprintf("tc filter add prio 66 dev %s egress bpf da obj %s sec classifier_egress", dev, obj))
-	err = cmd.Run()
-	if code := cmd.ProcessState.ExitCode(); code != 0 || err != nil {
-		return fmt.Errorf("failed to attach tc(egress) to %s, unexpected exit code: %d, err: %v", dev, code, err)
+	if err := rtnl.Filter().Add(&filter); err != nil {
+		return err
 	}
 
 	s.Lock()
 	s.qdiscs[inode] = qdisc{
 		netns:     netns,
 		device:    dev,
-		hasClsact: hasCls,
+		hasClsact: !find,
```

Comment:
this should be `find`?

Reply:
find means clsact already exists before we create, it was created by others.
hasClsact means that this clsact was created by us.
may be we should change `hasClsact` to `managedClsact` is better.